expressions/methods.md

```diff
@@ -29,14 +29,16 @@ After the return value there's a `=>` and then finally the function body. The va
 
 If you want to exit a function early then use the `return` command. If the function has a return type then you need to provide a value to return. If the function does not have a return type then `return` should appear on its own, without a value.
 
-```pony
-class C
-  fun factorial(x: U32): U32 =>
-    if x == 0 then
-	  return 1
-	end
+Pony applies tail call optimization to allow a recursive implementation such as the following factorial function:
 
-	x * factorial(x - 1)
+```pony
+fun factorial(x: I32): I32 ? =>
```

Comment:
If you're writing a function that needs to run recursively an arbitrary number of times, then it has to be tail call optimized – so I think it's important to note that you can and should rely on it in some situations.

It was mentioned on IRC and I thought it should go into the documentation that at least in this case, the optimization applies.

But it would be useful to have the compiler check it, e.g.
```pony
...
else
  inline fib(n - 1)
end
```